File: src/evai/command/SetCommands.java
package evai.command;

import mindustry.*;
import mindustry.world.*;
import mindustry.world.blocks.environment.*;
import mindustry.gen.*;
import mindustry.content.*;
import mindustry.world.consumers.*;
import mindustry.entities.units.*;
import mindustry.game.*;
import java.util.*;

public class SetCommands implements CommandRegistry.CommandHandler {
    @Override
    public CommandResult handle(String action, String[] args) {
        switch (action) {
            case "set": return set(args);
            case "nuke": return nuke(args);
            default: return new CommandResult(false, "unknown set command: " + action);
        }
    }

    private CommandResult set(String[] args) {
        if (args.length < 3) return new CommandResult(false, "usage: set [items/liquid/power/health/team/config/enabled/warmup/progress] <x> <y> <value>");
        int idx = 0;
        String sub = "";
        Set<String> subs = new HashSet<>(Arrays.asList("items", "liquid", "power", "health", "team", "config", "enabled", "warmup", "progress"));
        if (subs.contains(args[0])) { sub = args[0]; idx = 1; }
        Integer x = tryParseInt(args[idx], null);
        Integer y = tryParseInt(args[idx + 1], null);
        if (x == null || y == null) return new CommandResult(false, "invalid coordinates");
        Tile tile = Vars.world.tile(x, y);
        if (tile == null) return new CommandResult(false, "tile not found");
        Building bld = tile.build;
        if (!sub.isEmpty()) {
            String val = args.length > idx + 2 ? args[idx + 2] : "";
            return setSubField(bld, sub, val);
        }
        String field = args[idx + 2];
        String val = args.length > idx + 3 ? args[idx + 3] : "";
        return setReflectField(bld, field, val);
    }

    private CommandResult setSubField(Building bld, String sub, String value) {
        if (bld == null) return new CommandResult(false, "no building");
        try {
            switch (sub) {
                case "items": {
                    String[] parts = value.split(" ");
                    if (parts.length < 2) return new CommandResult(false, "usage: set items <x> <y> <item> <amount>");
                    mindustry.type.Item item = Vars.content.items().find(i -> i.name.equals(parts[0]));
                    if (item == null) return new CommandResult(false, "unknown item: " + parts[0]);
                    int amount = tryParseInt(parts[1], -1);
                    if (amount < 0) return new CommandResult(false, "invalid amount");
                    if (bld.items != null) bld.items.set(item, amount);
                    return new CommandResult(true, "Set " + parts[0] + " to " + amount);
                }
                case "liquid": {
                    String[] parts = value.split(" ");
                    if (parts.length < 2) return new CommandResult(false, "usage: set liquid <x> <y> <liquid> <amount>");
                    mindustry.type.Liquid liquid = Vars.content.liquids().find(l -> l.name.equals(parts[0]));
                    if (liquid == null) return new CommandResult(false, "unknown liquid: " + parts[0]);
                    float amount = tryParseFloat(parts[1], -1f);
                    if (amount < 0) return new CommandResult(false, "invalid amount");
                    if (bld.liquids != null) bld.liquids.set(liquid, amount);
                    return new CommandResult(true, "Set " + parts[0] + " to " + amount);
                }
                case "power": {
                    float amount = tryParseFloat(value, -1f);
                    if (amount < 0) return new CommandResult(false, "invalid amount");
                    float cap = 1f;
                    try { cap = bld.block.consPower.capacity; } catch (Exception ignored) {}
                    if (bld.power != null) bld.power.status = amount / cap;
                    return new CommandResult(true, "Set power to " + amount);
                }
                case "health": {
                    float h = tryParseFloat(value, -1f);
                    if (h < 0) return new CommandResult(false, "invalid amount");
                    bld.health(h);
                    return new CommandResult(true, "Set health to " + h);
                }
                case "team": {
                    mindustry.game.Team team = Team.all[0]; // fallback
                    for (Team t : Team.all) {
                        if (t != null && t.name.equalsIgnoreCase(value)) { team = t; break; }
                    }
                    bld.team(team);
                    return new CommandResult(true, "Set team to " + team.name);
                }
                case "config": {
                    bld.configure(value);
                    return new CommandResult(true, "Sent config");
                }
                case "enabled": {
                    bld.enabled = value.equals("on") || value.equals("true");
                    return new CommandResult(true, "Enabled: " + bld.enabled);
                }
                case "warmup": {
                    float w = tryParseFloat(value, -1f);
                    if (w < 0) return new CommandResult(false, "invalid warmup");
                    return setReflectField(bld, "warmup", value);
                }
                case "progress": {
                    float p = tryParseFloat(value, -1f);
                    if (p < 0) return new CommandResult(false, "invalid progress");
                    return setReflectField(bld, "progress", value);
                }
                default: return new CommandResult(false, "unknown sub: " + sub);
            }
        } catch (Exception e) {
            return new CommandResult(false, "error: " + e.getMessage());
        }
    }

    private CommandResult setReflectField(Building bld, String field, String value) {
        if (bld == null) return new CommandResult(false, "no building");
        try {
            Class<?> clazz = bld.getClass();
            while (clazz != null && clazz != Object.class) {
                try {
                    java.lang.reflect.Field f = clazz.getDeclaredField(field);
                    f.setAccessible(true);
                    Class<?> type = f.getType();
                    if (type == int.class || type == Integer.class) f.setInt(bld, Integer.parseInt(value));
                    else if (type == float.class || type == Float.class) f.setFloat(bld, Float.parseFloat(value));
                    else if (type == boolean.class || type == Boolean.class) f.setBoolean(bld, Boolean.parseBoolean(value));
                    else if (type == String.class) f.set(bld, value);
                    else if (type == double.class || type == Double.class) f.setDouble(bld, Double.parseDouble(value));
                    else f.set(bld, value);
                    return new CommandResult(true, "Set " + field + " = " + value);
                } catch (NoSuchFieldException e) {
                    clazz = clazz.getSuperclass();
                }
            }
            return new CommandResult(false, "field not found: " + field);
        } catch (Exception e) {
            return new CommandResult(false, "error: " + e.getMessage());
        }
    }

    private CommandResult nuke(String[] args) {
        Integer x = args.length > 0 ? tryParseInt(args[0], null) : null;
        Integer y = args.length > 1 ? tryParseInt(args[1], null) : null;
        if (x == null || y == null) return new CommandResult(false, "usage: nuke <x> <y> [radius]");
        int radius = args.length > 2 ? tryParseInt(args[2], 10) : 10;
        int count = 0;
        for (int dx = -radius; dx <= radius; dx++) {
            for (int dy = -radius; dy <= radius; dy++) {
                if (dx * dx + dy * dy > radius * radius) continue;
                Tile tile = Vars.world.tile(x + dx, y + dy);
                if (tile == null) continue;
                Building b = tile.build;
                if (b != null) { b.kill(); count++; }
            }
        }
        return new CommandResult(true, "Nuked " + count + " buildings");
    }

    private static Integer tryParseInt(String s, Integer fallback) {
        try { return Integer.parseInt(s); } catch (NumberFormatException e) { return fallback; }
    }
    private static int tryParseInt(String s, int fallback) {
        try { return Integer.parseInt(s); } catch (NumberFormatException e) { return fallback; }
    }
    private static float tryParseFloat(String s, float fallback) {
        try { return Float.parseFloat(s); } catch (NumberFormatException e) { return fallback; }
    }
}

File: src/evai/command/SetCommands.kt
package evai.command

import mindustry.*
import mindustry.gen.*
import mindustry.world.*
import evai.command.CommandRegistry.CommandHandler

class SetCommands : CommandHandler {
    override fun handle(action: String, args: Array<String>): CommandResult = when (action) {
        "set" -> {
            if (args.size < 3) return CommandResult(false, "usage: set [items/liquid/power/health/team/config/enabled/warmup/progress] <x> <y> <value>, or set <x> <y> <field> <value>")
            var idx = 0
            var sub = ""
            if (args[0] in listOf("items", "liquid", "power", "health", "team", "config", "enabled", "warmup", "progress")) {
                sub = args[0]
                idx = 1
            }
            val x = args[idx].toIntOrNull() ?: return CommandResult(false, "invalid x")
            val y = args[idx + 1].toIntOrNull() ?: return CommandResult(false, "invalid y")
            val tile = Vars.world.tile(x, y) ?: return CommandResult(false, "tile not found")
            val build = tile.build() ?: return CommandResult(false, "no building at $x,$y")
            if (sub.isNotEmpty()) {
                val valStr = if (args.size > idx + 2) args[idx + 2] else ""
                setSubField(build, sub, valStr)
            } else {
                val field = args[idx + 2]
                val valStr = if (args.size > idx + 3) args[idx + 3] else ""
                setReflectField(build, field, valStr)
            }
        }
        "nuke" -> {
            val x = args.getOrNull(0)?.toIntOrNull() ?: return CommandResult(false, "usage: nuke <x> <y> [radius]")
            val y = args.getOrNull(1)?.toIntOrNull() ?: return CommandResult(false, "usage: nuke <x> <y> [radius]")
            val radius = args.getOrNull(2)?.toIntOrNull() ?: 10
            for (dx in -radius..radius) {
                for (dy in -radius..radius) {
                    if (dx * dx + dy * dy > radius * radius) continue
                    val tile = Vars.world.tile(x + dx, y + dy) ?: continue
                    val build = tile.build() ?: continue
                    build.kill()
                }
            }
            CommandResult(true, "Nuked radius $radius at ($x, $y)")
        }
        else -> CommandResult(false, "unknown set command: $action")
    }

    private fun setSubField(build: Building, sub: String, value: String): CommandResult = try {
        when (sub) {
            "items" -> {
                val parts = value.split(" ")
                if (parts.size < 2) return CommandResult(false, "usage: set items <x> <y> <item> <amount>")
                val item = Vars.content.items().find { it.name == parts[0] } ?: return CommandResult(false, "unknown item: ${parts[0]}")
                val amount = parts[1].toIntOrNull() ?: return CommandResult(false, "invalid amount")
                build.items?.set(item, amount)
                CommandResult(true, "Set ${parts[0]} to $amount at building")
            }
            "liquid" -> {
                val parts = value.split(" ")
                if (parts.size < 2) return CommandResult(false, "usage: set liquid <x> <y> <liquid> <amount>")
                val liquid = Vars.content.liquids().find { it.name == parts[0] } ?: return CommandResult(false, "unknown liquid: ${parts[0]}")
                val amount = parts[1].toFloatOrNull() ?: return CommandResult(false, "invalid amount")
                build.liquids?.set(liquid, amount)
                CommandResult(true, "Set ${parts[0]} to $amount at building")
            }
            "power" -> {
                val amount = value.toFloatOrNull() ?: return CommandResult(false, "invalid amount")
                build.power?.status = amount / build.block.consumes?.power?.capacity?.toFloat() ?: 1f
                CommandResult(true, "Set power to $amount")
            }
            "health" -> {
                val amount = value.toFloatOrNull() ?: return CommandResult(false, "invalid amount")
                build.health = amount
                CommandResult(true, "Set health to $amount")
            }
            "team" -> {
                val team = mindustry.game.Team.all.find { it.name.equals(value, true) } ?: return CommandResult(false, "unknown team: $value")
                build.team = team
                CommandResult(true, "Set team to $value")
            }
            "config" -> {
                build.configure(value)
                CommandResult(true, "Sent config: $value")
            }
            "enabled" -> {
                val en = value == "on" || value == "true"
                build.enabled = en
                CommandResult(true, "Enabled: $en")
            }
            "warmup" -> {
                val w = value.toFloatOrNull() ?: return CommandResult(false, "invalid warmup (0-1)")
                build.warmup = w.coerceIn(0f, 1f)
                CommandResult(true, "Warmup: $w")
            }
            "progress" -> {
                val p = value.toFloatOrNull() ?: return CommandResult(false, "invalid progress")
                build.progress = p
                CommandResult(true, "Progress: $p")
            }
            else -> CommandResult(false, "unknown sub: $sub")
        }
    } catch (e: Exception) {
        CommandResult(false, "error: ${e.message}")
    }

    private fun setReflectField(build: Building, field: String, value: String): CommandResult = try {
        var clazz: Class<*> = build.javaClass
        var found = false
        while (clazz != Any::class.java) {
            try {
                val f = clazz.getDeclaredField(field)
                f.isAccessible = true
                val type = f.type
                when {
                    type == Int::class.java || type == Int::class.javaPrimitiveType -> f.setInt(build, value.toInt())
                    type == Float::class.java || type == Float::class.javaPrimitiveType -> f.setFloat(build, value.toFloat())
                    type == Boolean::class.java || type == Boolean::class.javaPrimitiveType -> f.setBoolean(build, value.toBoolean())
                    type == String::class.java -> f.set(build, value)
                    type == Double::class.java || type == Double::class.javaPrimitiveType -> f.setDouble(build, value.toDouble())
                    type == Long::class.java || type == Long::class.javaPrimitiveType -> f.setLong(build, value.toLong())
                    else -> f.set(build, value)
                }
                found = true
                break
            } catch (_: NoSuchFieldException) {
                clazz = clazz.superclass
            }
        }
        if (found) CommandResult(true, "Set $field = $value")
        else CommandResult(false, "field not found: $field")
    } catch (e: Exception) {
        CommandResult(false, "error setting $field: ${e.message}")
    }
}